zarr/core.py

```diff
@@ -1980,9 +2008,29 @@ def _chunk_getitems(self, lchunk_coords, lchunk_selection, out, lout_selection,
                 for ckey in ckeys
                 if ckey in self.chunk_store
             }
+        elif (
+            self._partial_decompress
+            and not self._compressor
+            and not fields
```

Comment:
fields *are* allowed, and the final dtype might be object if one of the intermediate codecs turns bytes into objects (e.g., categorical). Here we are asserting that any compression must change the number of bytes, which might not be true.

Reply:
@jstriebel, any comment here?

Reply:
> fields _are_ allowed

I think that's right, we might not need to exclude `fields` usage here. However, I'm wondering the same for `PartialReadBuffer`, why is it excluded there? I assumed the same reason applies now.

> the final dtype might be object if one of the intermediate codecs turns bytes into objects (e.g., categorical).

True, but is there any option to recognize that? Again, this is the same issue as with `PartialReadBuffer` above.

> Here we are asserting that any compression must change the number of bytes, which might not be true.

We use it as an exclusion criterum, yes, because most compressions do and again, there's currently no way to tell if it would change the number of bytes. Again, this is exactly the same as above with `PartialReadBuffer`.

I agree that there might be more cases where `PartialReadBuffer` could be used, but this solves it for the ones where I'm sure that it works.

Reply:
@martindurant: how do you see moving forward here? Do you think these issues (/changes in assumptions) can impact existing code?

Reply:
Existing code, no.

But we want to make something that is as generally useful as possible. The points here suggest to me that such decisions are poorly made at the array layer, and something only the codecs (including compressors) can know, if they are fed the information about what the request was.

Reply:
Perhaps this should be improved in a follow-up PR, I don't think this should block the current one. If anything, this is purely improving the current state since it allows partial reads for uncompressed chunks at all.

Reply:
Thanks, both. Had this on my radar for discussion at the meeting and missed that it had happened on GitHub 🤦🏽